src/ipips/ipip-0421.md

```diff
@@ -0,0 +1,115 @@
+- Defines Hashing and Encryption functions and response payloads structure.
+
+There are no ideomatic changes to the API - all data formats, design rationale and principles outlined in the original [HTTP Delegated Routing IPIP](./ipip-0337.md) apply here. 
+
+### User benefit
+
+With the new APIs users can protect themselves from:
+- a malicious actor spying on the user by observing the user to Content Router traffic and then downloading the same data;
+- the new API is a first step towards fully private HTTP Delegated Routing protocol that will eliminate IPNI as centralised observers.
+
+There are no other functional improvements.
+
+### Compatibility
+
+#### Backwards Compatibility
+
+Users will need to explicitly turn on Reader Privacy on their nodes. A new flag can be introduced to the Kubo's HTTP Delegated Content Router configuration to facilitate that functionality. 
+Users on older nodes can continue using the old API and turn on reader Privacy at a alter point. 
+
+Content Routers should provide the same QoS for both Privacy Preserving and regular APIs. This is because both can be served over the same encrypted data. If presented with a regular CID, a Content Router 
+can perform decryption operations on behalf of the user (i.e. mimic the client logic) and return results in clear text. If presented with a second hash the Content Router can return encrypted results and let the
+user to do decryption themselves.
+
+It's possible that not all Content Routers will adopt Reader Privacy. The default HTTP Delegated Router like `cid.contact` should have Reader Privacy enabled by default in the newer versions of Kubo / Helia. 
+Users should verify themselves whether a custom router of their choice supports Reader Privacy or not when configuring it.
+
+The `/routing/v1/encrypted/` API will be implemented in [`boxo/routing/http`](https://github.com/ipfs/boxo/tree/main/routing/http) and will not introduce any breaking changes to existing clear text endpoints.
+The API will be released in a new minor version. 
+
+#### Forwards Compatibility
+
+Reader Privacy relies on usage of specific hashing and encryption functions. Function rotation will require a network-wide migration. Content Routers might not be able to migrate "under the hood" as they
+don't possess the original values. Function rotation should be a very infrequent event and will require network-wide efforts. When function rotation is needed - a version of the API will be incremented.  
+
+### User benefit
+
+TODO: How will end users benefit from this work?
+
+### Security
+
+TODO:  Explain the security implications/considerations relevant to the proposed change.
+
+### Alternatives
+
+TODO: Describe alternate designs that were considered and related work.
+
+- TODO:  Oblivious HTTP ([IETF](https://www.ietf.org/archive/id/draft-thomson-http-oblivious-01.html), [Cloudflare](https://blog.cloudflare.com/stronger-than-a-promise-proving-oblivious-http-privacy-properties/))
```

Comment:
I am trying to wrap my head around threat models for two hypothetical worlds (those are not exclusive):
- (A) IPNI `/routing/v1` supporting Double Hashed records (this IPIP)
- (B) IPNI `/routing/v1` behind OHTTP ([IETF](https://www.ietf.org/archive/id/draft-thomson-http-oblivious-01.html), [Cloudflare](https://blog.cloudflare.com/stronger-than-a-promise-proving-oblivious-http-privacy-properties/))

My understanding is that in (A)  user needs to trust IPNI operator anyway to some degree, because IPNI's HTTP endpoint still can track a bunch of metadata about users query usage patterns, hours they are online, their IPs, user-agent, etc.

I think it would be beneficial to the reader privacy story if "Alternatives" section here mentioned OHTTP or Onion Services as a complimentary solutions that can be implemented and deployed in addition to this IPIP close the described gap.

Reply:
> Onion Services as a complimentary solutions


Agreed. I will take this.